File: algorithms/2_medium/M/01647_minimum_deletions_to_make_character_frequencies_unique/solution.c
static int cmp(const void *a_ptr, const void *b_ptr) {
    const size_t a = *(size_t*)a_ptr;
    const size_t b = *(size_t*)b_ptr;

    if (a < b) {
        return 1;
    } else if (a == b) {
        return 0;
    } else {
        return -1;
    }
}

int minDeletions(const char *restrict s) {
    if (s == NULL || s[0] == '\0') { return 0; }

    size_t freqs[256] = { 0 };
    for (size_t i = 0; s[i] != '\0'; i += 1) {
        freqs[(unsigned char)s[i]] += 1;
    }
    qsort(freqs, 256, sizeof(size_t), cmp);
    size_t cnt = 0;
    size_t prev = freqs[0];
    for (size_t i = 1; i < 256; i += 1) {
        if (freqs[i] == 0) {
            break;
        }

        if (prev > 0) {
            prev -= 1;
        }

        if (freqs[i] > prev) {
            cnt += freqs[i] - prev;
        }

        prev = MIN(prev, freqs[i]);
    }

    if (cnt > (size_t)INT_MAX) { return -1; }

    return (int)cnt;
}

File: algorithms/2_medium/M/01647_minimum_deletions_to_make_character_frequencies_unique/solution.rs
impl Solution {
    pub fn min_deletions(s: String) -> i32 {
        let mut freqs = [0_usize; 256];
        for &b in s.as_bytes() {
            freqs[b as usize] += 1;
        }
        freqs.sort_unstable_by(|a, b| b.cmp(&a));
        let mut cnt = 0_usize;
        let mut prev = freqs[0];
        for f in freqs.into_iter().filter(|&x| x > 0).skip(1) {
            if prev > 0 {
                prev -= 1;
            }

            if f > prev {
                cnt += f - prev;
            }

            prev = prev.min(f);
        }

        cnt.try_into().unwrap()
    }
}
